kore/src/Kore/Builtin/AssocComm/AssocComm.hs

```diff
@@ -103,3 +114,66 @@ splitLastInit [a] = Just ([], a)
 splitLastInit (a : as) = do
     (initA, lastA) <- splitLastInit as
     return (a : initA, lastA)
+
+type SyntaxPattern variable = Recursive.Base
+    (Syntax.Pattern variable Attribute.Null)
+    (Free
+        (Recursive.Base (Syntax.Pattern variable Attribute.Null))
+        (TermLike variable)
+    )
+
+externalize1 ::
+    forall variable.
+    InternalVariable variable =>
+    UnitSymbol ->
+    ConcatSymbol ->
+    [SyntaxPattern variable] ->
+    [SyntaxPattern variable] ->
+    [TermLike variable] ->
+    Either (TermLike variable) (SyntaxPattern variable)
+externalize1
+    (UnitSymbol unitSymbol)
+    (ConcatSymbol concatSymbol)
+    concreteElements
+    variableElements
+    opaque =
+        case splitLastInit concreteElements of
+            Nothing -> case splitLastInit opaque of
+                Nothing -> case splitLastInit variableElements of
+                    Nothing -> pure unit
+                    Just (ves, ve) -> pure $ addTerms ves ve
+                Just (opaqs', opaq1) -> case splitLastInit opaqs' of
+                    Nothing -> case splitLastInit variableElements of
+                        Nothing -> Left opaq1
+                        Just (ves, ve) -> pure $
+                            addTerms ves $
+                                concatOpaques' opaq1 ve
+                    Just (opaqs, opaq2) -> pure $
+                        addTerms variableElements $
+                            addTermsOpaque opaqs $
+                                twoOpaques opaq1 opaq2
+            Just (concretes, concrete) -> pure $
+                addTerms variableElements $
+                    addTermsOpaque opaque $
+                        addTerms concretes concrete
```

Comment:
We could clarify this in this way:

```suggestion
externalize1
    (UnitSymbol unitSymbol)
    (ConcatSymbol concatSymbol) = worker
  where
    worker [] [] [] = unit
    worker [] [] [singleOpaque] = Left singleOpaque
    worker concreteElements symbolicElements opaqueTerms =
        Right $ foldr1 concat' operands
      where
        operands = (map Free concreteElements) <> (map Free symbolicElements) <> (map Pure operands)
```

Another way might be:

```suggestion
externalize1
    (UnitSymbol unitSymbol)
    (ConcatSymbol concatSymbol) = worker
  where
    worker [] [] [] = unit
    worker concreteElements symbolicElements opaqueTerms =
        Right $ foldr concat' unit operands
      where
        operands = (map Free concreteElements) <> (map Free symbolicElements) <> (map Pure operands)
```

Reply:
I think the first suggestion is not type correct.
the second suggestion might work (and we won't even need the base `[] [] []` case) but has the disadvantage that it always prepends a unit at the end, but that's probably fine if all we do is unparse.

Reply:
I edited my suggestion above to remove the one that appends the unit (we decided that was wrong) and fixes the version using `foldr1`.